packages/muelu/src/Graph/MatrixTransformation/MueLu_BoundaryDetection.hpp

```diff
@@ -0,0 +1,243 @@
+    , functor3(0) {}
+
+  BoundaryFunctor(local_matrix_type& A_, functor_type_0& functor0_, functor_type_1& functor1_, functor_type_2& functor2_, functor_type_3& functor3_)
+    : A(A_)
+    , functor0(functor0_)
+    , functor1(functor1_)
+    , functor2(functor2_)
+    , functor3(functor3_) {}
+
+  KOKKOS_INLINE_FUNCTION
+  void
+  operator()(const local_ordinal_type rlid) const {
+    if constexpr (!std::is_same_v<functor_type_0, int>)
+      functor0(rlid);
+    if constexpr (!std::is_same_v<functor_type_1, int>)
+      functor1(rlid);
+    if constexpr (!std::is_same_v<functor_type_2, int>)
+      functor2(rlid);
+    if constexpr (!std::is_same_v<functor_type_3, int>)
+      functor3(rlid);
+  }
+};
+
+// Marks rows as Dirichlet based on value threshold and number of off-diagonal entries.
+// Marks blocks as Dirichlet when one row is Dirichlet (useGreedyDirichlet==true) or when all rows are Dirichlet (useGreedyDirichlet==false).
+template <class local_matrix_type, bool useGreedyDirichlet>
+class VectorDirichletFunctor {
+ private:
+  using scalar_type        = typename local_matrix_type::value_type;
+  using local_ordinal_type = typename local_matrix_type::ordinal_type;
+  using memory_space       = typename local_matrix_type::memory_space;
+
+  using ATS                 = Kokkos::ArithTraits<scalar_type>;
+  using magnitudeType       = typename ATS::magnitudeType;
+  using boundary_nodes_view = Kokkos::View<bool*, memory_space>;
+
+  local_matrix_type A;
+  local_ordinal_type blockSize;
+  boundary_nodes_view boundaryNodes;
+  magnitudeType dirichletThreshold;
+  local_ordinal_type dirichletNonzeroThreshold;
+
+ public:
+  VectorDirichletFunctor(local_matrix_type& A_, local_ordinal_type blockSize_, boundary_nodes_view boundaryNodes_, magnitudeType dirichletThreshold_, local_ordinal_type dirichletNonzeroThreshold_)
+    : A(A_)
+    , blockSize(blockSize_)
+    , boundaryNodes(boundaryNodes_)
+    , dirichletThreshold(dirichletThreshold_)
+    , dirichletNonzeroThreshold(dirichletNonzeroThreshold_) {}
+
+  KOKKOS_FORCEINLINE_FUNCTION
+  void operator()(const local_ordinal_type rblid) const {
+    for (local_ordinal_type rlid = rblid * blockSize; rlid < (rblid + 1) * blockSize; ++rlid) {
+      auto row               = A.rowConst(rlid);
+      local_ordinal_type nnz = 0;
+      bool rowIsDirichlet    = true;
+      for (local_ordinal_type k = 0; k < row.length; ++k) {
+        auto clid = row.colidx(k);
+        auto val  = row.value(k);
+        if ((rlid != static_cast<local_ordinal_type>(clid)) && (ATS::magnitude(val) > dirichletThreshold)) {
+          ++nnz;
+          if (nnz == dirichletNonzeroThreshold) {
+            rowIsDirichlet = false;
+            break;
+          }
+        }
+      }
+      if constexpr (useGreedyDirichlet) {
+        if (rowIsDirichlet) {
+          boundaryNodes(rblid) = true;
+          return;
+        }
+      } else {
+        if (!rowIsDirichlet) {
+          boundaryNodes(rblid) = false;
```

Comment:
MAybe remove?

Reply:
`boundaryNodes` is initialized to all `false`. Then we apply one or more functors that can switch entries to `true`. We probably don't want to override previous decisions to mark a node as Dirichlet. Currently we don't use more than one functor, but that seems reasonable as a design decision, right?